Feature app launcher

peyman-mohtashami opened the pull request:
Add the "App Launcher" question type to questions.
Create a field in a questionnaire with the type of `launcher` and `add app_name`, `android_package_name` and `ios_package_name`:
```json
{
    "field_type": "launcher",
    "app_name": "...",
    "android_package_name": "...",
    "ios_package_name": "...",
    "field_label": "...",
    ...
},
```

blootsvoets commented on src/app/pages/questions/components/question/app-launcher/app-launcher.component.ts:
```diff
@@ -0,0 +1,71 @@
+      options.uri = this.iosPackageName
+    } else {
+      // eu.dynamore.app://dynamore/questionnaire_completed?timestamp= 1612452411000
+      options.packageName = this.androidPackageName + '://dynamore/questionnaire_completed?timestamp=' + this.timestamp
```
To invoke a deep link, `options.uri` should be used instead. The questionnaire configuration should include that url, excluding the timestamp query parameter. Then the uri can be built as follows:
```typescript
options.uri = this.androidUri
if ('?' in options.uri) {
  options.uri += '&'
} else {
  options.uri += '?'
}
options.uri += 'timestamp=' + this.timestamp
```

blootsvoets reviewed (commented):
Some comment inline. After reconsideration, would it be possible to make the app launcher task part of the finish-questionnaire flow? The flow would then be:

Button [ Finish questionnaire and start other app ] -> finalise/send questionnaire -> launch other app.

This way, the user does not need to return to the current app to finish questionnaire. Since they are likely to forget that. If the user does not finish the questionnaire, it will remain as an open task and the results will not be sent to the backend.

blootsvoets commented:
Adding in @hwesley for reference of the feature.

Another consideration: if no iOS link is given, no action should be taken in iOS, and vice versa for Android. Perhaps in general, if the link cannot be invoked, show some text that the action cannot be taken?

peyman-mohtashami commented on src/app/pages/questions/containers/questions-page.component.ts:
```diff
@@ -121,10 +134,42 @@ export class QuestionsPageComponent implements OnInit {
+    }
+    options.uri += 'timestamp=' + this.task.timestamp
+
+    this.appLauncher.canLaunch(options)
+      .then((canLaunch: boolean) => {
+        if(canLaunch){
+          this.appLauncher.launch(options).then(()=>{
+            console.log('App launched')
+          }, (err)=>{
+            console.log('Error in launching app', err)
+          })
+        } else {
+          console.log('App cannot be launched')
+        }
+      })
+      .catch((error: any) => console.log('Error in checking if the app can launch', error))
```
@blootsvoets I added some logs instead of showing the user that the uri is not reachable in the device. If the uri is not reachable or there is an error in opening the app, nothing happens.

blootsvoets replied:
Would it be possible instead to already call the `canLaunch()` when initialising so the text can be changed to show the user does not support the package? Then the console can log the error if the `launch()` function fails.

peyman-mohtashami replied:
It's possible but the problem is how the user knows there is something remained after finishing. 
In this scenario, I think we should tell the user that if you press `finish button` an external app will open. 
1) Do you have any suggestion?
And there, if the app doesn't exist on the device (the uri is not reachable) there should be an alert.
2) I can set an alert as a toast or show alert from `alertService`. Which one do you think is better?

blootsvoets replied:
Indeed, I think if the another app will be launched after finish, the button should show `Launch ...` or `Launch to finish` and the text should say this will launch ... app and finish the questionnaire.

As for alerts, I'd think a dedicated page would be more intuitive for people, but I'm not sure if there's a dedicated page after the finish page? If not, an alert on the home screen would be enough I suppose.

peyman-mohtashami reviewed (commented):
I changed it a little:
`android_package_name` -> `android_uri`
`ios_package_name` -> `ios_uri`

Move the app-launcher code to finish flow before navigating to home page

Modify `options.package` to `options.uri`

peyman-mohtashami commented on src/assets/data/localisations.ts:
```diff
@@ -176,6 +176,46 @@ export const Localisations = {
     pl: 'Dziękujemy za wypełnienie ankiety.',
     hb: '.תודה שמילאת את השאלון'
   },
+  LAUNCH_APP_DESC: {
+    da: '',
+    de: '',
+    en: 'Finish the questionnaire and launch',
+    es: '',
+    it: '',
+    nl: '',
+    pl: '',
+    hb: ''
+  },
```
The translations should be completed.

peyman-mohtashami reviewed (commented):
I added an alert if the app cannot launch and put it after the user presses the finish/launch app button.
Do you mean that?

peyman-mohtashami commented:
I refactored the code and create a service and component for app launcher.
Also changed the names of attributes in questions to be more meaningful:
```
"field_type": "launcher",
"external_app_android_uri": "",
"external_app_android_uri": "",
"external_app_name": "", 
```

blootsvoets commented on src/app/core/services/config/remote-config.service.ts:
```diff
@@ -130,7 +130,17 @@ export class FirebaseRemoteConfigService extends RemoteConfigService {
+    const date = Date.now();
+    let currentDate = null;
+    do {
+      currentDate = Date.now();
```
This is a busy-waiting sleep, really not recommended. Prefer to use actual timeouts.

blootsvoets commented on src/assets/data/defaultConfig.ts:
```diff
@@ -47,8 +47,10 @@ export const DefaultOnDemandAssessmentLabel = JSON.stringify(
+//   'https://radar-cns-platform.rosalind.kcl.ac.uk'
 export const DefaultEndPoint: string =
-  'https://radar-cns-platform.rosalind.kcl.ac.uk'
+  'https://radar-test.thehyve.net'
```
This should be reverted.

blootsvoets commented on src/assets/data/defaultConfig.ts:
```diff
@@ -131,7 +133,8 @@ export const DefaultNotificationTtlMinutes: number = 10
 // *The Github repository where the protocols are located (REMOTE CONFIG KEY: `protocol_repo`)
-export const DefaultProtocolGithubRepo = 'RADAR-Base/RADAR-aRMT-protocols'
+// export const DefaultProtocolGithubRepo = 'RADAR-Base/RADAR-aRMT-protocols'
+export const DefaultProtocolGithubRepo = 'peyman-mohtashami/RADAR-aRMT-protocols'
```
This should be reverted

blootsvoets commented on src/app/pages/questions/services/app-launcher.service.ts:
```diff
@@ -0,0 +1,124 @@
+      .then((canLaunch: boolean) => {
+        return canLaunch;
+      })
+      .catch(() => {
+        return false
```
This should probably be logged with the loggerservice?

blootsvoets reviewed (commented):
Mostly looks good to me. Before it can be merged into the general app, some adjustments need to be made.

blootsvoets commented:
@mpgxvii @yatharthranjan feature I just mentioned.
